module/yzecoriolis.js

```diff
@@ -387,16 +387,18 @@ Hooks.on("getSceneControlButtons", (controls) => {
 
 Hooks.once("ready", async function () {
   // Determine whether a system migration is required and feasible
-  const currentVersion = game.settings.get(
+  const currentVersion = game.system.version;
+  const lastMigratedToVersion = game.settings.get(
     "yzecoriolis",
     "systemMigrationVersion"
   );
-  const NEEDS_MIGRATION_VERSION = "2.5.0";
+
+  const NEEDS_MIGRATION_AFTER_VERSION = "3.0.0";
   const COMPATIBLE_MIGRATION_VERSION = "1.4.7";
+
   let needMigration =
-    currentVersion &&
-    isNewerVersion(NEEDS_MIGRATION_VERSION, currentVersion) &&
-    currentVersion !== "0"; // zero means we just created a new world.
+    isNewerVersion(currentVersion, NEEDS_MIGRATION_AFTER_VERSION) &&
+    isNewerVersion(NEEDS_MIGRATION_AFTER_VERSION, lastMigratedToVersion);
```

Comment:
Modified a bit the code here. There was an issue for system that never had any migration run... because the `systemMigrationVersion` is only set on a successful migration, any world created with a version after the last migration version, would just have 0 set. Consequently, `needMigration` was always false because of line 399.

So now I use both the current system version and the last migrated version to determine if a migration is needed.

A migration is needed if:
- the currentVersion is newer than the version after which we need to migrate (3.1.0) in this case
- AND version after which we need to migrate is newer that the last migrated version (store in the settings)

Reply:
@coderunner great catch! Thank you!